types/tls.go

```diff
@@ -1,43 +1,41 @@
 package types
 
 import (
-	"encoding/json"
 	"fmt"
 )
 
-// TLSVersion is a strongly-typed int used for relative comparison
-// (minimum, maximum, greater than, less than) of TLS versions
-type TLSVersion int
+// TLSVersion is a strongly-typed string for TLS versions
+type TLSVersion string
 
 const (
 	// Error value, excluded from lookup maps
-	TLSVersionInvalid TLSVersion = iota - 1
+	TLSVersionInvalid TLSVersion = "TLS_INVALID"
 
 	// Explicit unspecified zero-value to avoid overwriting parent defaults
-	TLSVersionUnspecified
+	TLSVersionUnspecified TLSVersion = ""
 
 	// Explictly allow implementation to select TLS version
 	// May be useful to supercede defaults specified at a higher layer
-	TLSVersionAuto
+	TLSVersionAuto TLSVersion = "TLS_AUTO"
 
 	_ // Placeholder for SSLv3, hopefully we won't have to add this
 
 	// TLS versions
-	TLSv1_0
-	TLSv1_1
-	TLSv1_2
-	TLSv1_3
+	TLSv1_0 TLSVersion = "TLSv1_0"
+	TLSv1_1 TLSVersion = "TLSv1_1"
+	TLSv1_2 TLSVersion = "TLSv1_2"
+	TLSv1_3 TLSVersion = "TLSv1_3"
 )
 
 var (
-	TLSVersions = map[string]TLSVersion{
+	tlsVersions = map[string]TLSVersion{
```

Comment:
Is this map just for verifying the "enum"? You could just do it this way instead then:
```
tlsVersions = map[string]struct{}{
    TLSVersionAuto: struct{}{},
    TLSv1_0: struct{}{}, 
    // etc
}
```
Then later below you can just use `_, ok := tlsVersions[v]` to check for validity.

Reply:
`TLS_AUTO` mapping to `TLSVersionAuto` is kinda the outlier here (it could be changed), but the general intent was to have an explicit lookup mapping between external config and internal values, particularly given the "old" Consul client TLS config which will be updated in https://github.com/hashicorp/consul/pull/11647

Reply:
Okay, implemented this but now I'm thinking it feels quite strange to have a validation function which takes `TLSVersion` as an argument since that would imply the (now `string` type) can contain invalid data anywhere unless it's validated explicitly, which I guess is true. I just want real enums or at least a constructor to keep garbage data out 😞